### PR TITLE
1729 Writers Proof of Contribution SBT implementation

### PR DESCRIPTION
**Pull request by neodaoist**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,4 +3,5 @@ out/
 target/
 node_modules/
 Cargo.lock
-
+.vscode/
+script/
```
Comment (plaird523) [inline .gitignore:7]: Did you mean to add script/ to .gitignore?  We already have tracked files there that seem useful to have

File: src/ISoulbound.sol [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    /*//////////////////////////////////////////////////////////////
+                        Events
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Explain to an end user what this does
+    /// @dev Explain to a developer any extra details
+    /// @param _issuer Documents a parameter just like in doxygen (must be followed by parameter name)
+    /// @param _issuee Documents a parameter just like in doxygen (must be followed by parameter name)
+    /// @param _tokenId Documents a parameter just like in doxygen (must be followed by parameter name)
```
Comment (plaird523) [inline src/ISoulbound.sol:19]: Boilerplate doc -- should add real doc or remove

File: src/ISoulbound.sol [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        address indexed _issuee,
+        uint256 _tokenId
+    );
+
+    /// @notice Explain to an end user what this does
+    /// @dev Explain to a developer any extra details
+    /// @param _revoker Documents a parameter just like in doxygen (must be followed by parameter name)
+    /// @param _revokee Documents a parameter just like in doxygen (must be followed by parameter name)
+    /// @param _tokenId Documents a parameter just like in doxygen (must be followed by parameter name)
+    /// @param _reason Documents a parameter just like in doxygen (must be followed by parameter name)
```
Comment (plaird523) [inline src/ISoulbound.sol:31]: Boilerplate doc (there's a lot of it in this file, I wont keep duplicating comments)

File: src/SBT.sol [hunk 4]
```diff
@@ -1,88 +1,90 @@
...
+    /// @inheritdoc	ISoulbound
+    function issue(address _recipient, uint256 _tokenId) external onlyOwner {
+        _issue(_recipient, _tokenId);
+    }
 
-    function issueBatch(
-        address[] calldata _recipients,
-        uint256[] calldata _tokenIDs
-    ) external;
+    function _issue(address _recipient, uint256 _tokenId) internal {
```
Comment (plaird523) [inline src/SBT.sol:58]: The pattern of _issue doing the actual work for issue and issueBatch makes sense but was non-obvious at first.  Some doc would be helpful:
issue -- wrapper to issue a single token
issueBatch -- wrapper to issue a batch of tokens to a list of recipients
_issue -- logic to issue a single token, as part of individual issue call or batch

File: test/SBT.t.sol [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+    /*//////////////////////////////////////////////////////////////
+                        Issuing
+    //////////////////////////////////////////////////////////////*/
+
+    // TODO decide if we should enforce 1 credential of each type per user or support multiple / max
+    // if yes — replace amount of 1 in SBT with something other than a magic number
+    // if no — add to functions and events, consider adding an immutable constructor arg for max
+
+    // TODO decide if we should support ownership transfer in SBT constructor, or leave to implementers
```
Comment (plaird523) [inline test/SBT.t.sol:23]: Suggestion: do it in the constructor OR provide a concrete implementation that does it in its constructor

**Review by plaird523 — CHANGES_REQUESTED** [review]
A few comments in-line.  Generally needs doc.

Also, how do the token IDs work here?  Do we need to implement a counter or will they have a significance and end up being reused (like in batches)?

File: test/SBT.t.sol (hunk 5 above)
Comment (neodaoist) [reply test/SBT.t.sol:23]: FWIW OZ Ownable already does this, for the deploying EOA. I will add to our 1729 Writers contract, not the base SBT contract.

File: .gitignore (hunk 1 above)
Comment (neodaoist) [reply .gitignore:7]: Good catch, reverting

**Review by plaird523 — APPROVED** [review]

**Pull request merged**